routing/pathfind_test.go

```diff
@@ -1290,6 +1293,54 @@ func runPathFindingWithAdditionalEdges(t *testing.T, useCache bool) {
 	assertExpectedPath(t, graph.aliasMap, path, "songoku", "doge")
 }
 
+// runPathFindingWithRedundantAdditionalEdges asserts that we are able to
+// find paths to nodes ignoring additional edges that are already known by
+// self node.
+func runPathFindingWithRedundantAdditionalEdges(t *testing.T, useCache bool) {
+	graph, err := parseTestGraph(t, useCache, basicGraphFilePath)
+	require.NoError(t, err, "unable to create graph")
+
+	sourceNode, err := graph.graph.SourceNode()
+	require.NoError(t, err, "unable to fetch source node")
+
+	paymentAmt := lnwire.NewMSatFromSatoshis(100)
+
+	// Create the channel edge going from roasbeef to songoku and include it in
+	// our map of additional edges.
+	var hintChannelId uint64 = 1337
+	roasbeefToSongoku := &channeldb.CachedEdgePolicy{
+		ToNodePubKey: func() route.Vertex {
+			return graph.aliasMap["songoku"]
+		},
+		ToNodeFeatures:            lnwire.EmptyFeatureVector(),
+		ChannelID:                 hintChannelId,
+		FeeBaseMSat:               1,
+		FeeProportionalMillionths: 1000,
+		TimeLockDelta:             9,
+	}
+
+	additionalEdges := map[route.Vertex][]*channeldb.CachedEdgePolicy{
+		graph.aliasMap["roasbeef"]: {roasbeefToSongoku},
+	}
+
+	// We should now be able to find a path from roasbeef to doge.
+	path, err := dbFindPath(
+		graph.graph, additionalEdges, &mockBandwidthHints{},
+		noRestrictions, testPathFindingConfig,
+		sourceNode.PubKeyBytes, graph.aliasMap["songoku"], paymentAmt,
+		0, 0,
+	)
+
+	require.NoError(t, err, "unable to find path to songoku")
+
+	// The path should represent the following hop roasbeef -> songoku
+	assertExpectedPath(t, graph.aliasMap, path, "songoku")
+
+	if path[0].ChannelID == hintChannelId {
+		t.Fatal("additional edge for known edge wasn't ignored")
+	}
+}
```

Comment:
Would it be worth to add a test on a higher level that asserts that the infinite loop is no longer occuring?

Reply:
Definitely would, do you have a suggestion for where? We'll get right on it

Reply:
Perhaps `integrated_routing_test.go` is suitable for this. Not as heavy as an integration test, but it does cover the various interacting components for the bug that is fixed in this PR.

Reply:
[92c372a](https://github.com/lightningnetwork/lnd/pull/6766/commits/92c372aad2b0919655d62bf0f9041776823936a1)

We gave it a go and ran into some difficulties. The bigger one was that when hop hints are passed the target public key is actually parsed, which means it cannot be an invalid set of bytes. We made it work improving the way the test creates the Vertex by actually turning the input byte into a valid public key, as you'll see from the commit.

Without the change we did, the test we coded fails due to unknown next peer, but it doesn't go into a loop. The loop probably happens on the lifecycle or some level above, but with our change what would be the unknown next peer actually turns into a successful payment (as tested), which is even better for this case.

What do you think? Should we test on some other level or are the already submitted tests enough?

Reply:
The test does look good, but I am surprised that without the fix the infinite loop isn't reproduced. It does use payment session and mission control, so would expect that `RequestRoute` to return the same result each time. Maybe the bad route hint should be bad in a different way?

A definitive answer to this question would be great to have, so that there is certainty on what is covered exactly. I don't think it will be needed to write a test on yet another level.

Reply:
@joostjager Makes sense, me and @andreihod will investigate further and try to understand this behavior. We were able to reproduce the loop locally and verify that this fix works, so we'll try to explain why the test case is different from our testing.

Reply:
@joostjager Upon more investigation, we came to the conclusion that the loop is not reproduced in the test added to `integrated_routing_test` because it has it's own lifecycle loop (https://github.com/lightningnetwork/lnd/blob/master/routing/integrated_routing_context_test.go#L211) instead of using the real one (https://github.com/lightningnetwork/lnd/blob/master/routing/payment_lifecycle.go#L156), which causes the loop because the unknown next peer on first hop leads to a retry. We thought of adding one more test to `router_test` that triggers the loop exactly, but we weren't sure it was needed from what you mentioned before. How would you recommend we proceed?

Reply:
@joostjager [32db8f3](https://github.com/lightningnetwork/lnd/pull/6766/commits/32db8f36eadd1d4dd714dc42616ad0b1a0709487) Me and @andreihod ended up adding the missing test that exactly reproduces the infinite loop without the fix. We think we have all that we need, can you think of anything left?

Reply:
Nice investigation. Is https://github.com/lightningnetwork/lnd/blob/202fcd6672a004881925710e6dc9e336af8945aa/routing/mock_graph_test.go#L68 the difference that causes the different behavior in the test - not returning a real unknown next peer error?

The `router_test.go` addition looks good. Probably better than using the integrated routing test, because it has more fake components.

Now that you have both tests, you could leave them to see what a Lightning Labs reviewer prefers?